src/librustc_errors/emitter.rs

```diff
@@ -785,12 +792,72 @@ fn stderr_isatty() -> bool {
 
 pub enum Destination {
     Terminal(Box<term::StderrTerminal>),
+    BufferedTerminal(Box<BufferedStderr>),
     Raw(Box<Write + Send>),
 }
 
+pub struct BufferedWriter {
+    buffer: Vec<u8>,
+}
+
+impl BufferedWriter {
+    fn _new() -> BufferedWriter {
+        BufferedWriter {
+            buffer: vec![]
+        }
+    }
+}
+
+impl Write for BufferedWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        for b in buf {
+            self.buffer.push(*b);
+        }
+        Ok(buf.len())
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        let mut stderr = io::stderr();
+        if let Ok(buff) = String::from_utf8(self.buffer.clone()) {
+            write!(stderr, "{}", buff)?;
+        }
```

Comment:
Instead of converting to utf-8 this could just use `io::stderr().write_all(&self.buffer)` perhaps?